pkg/runtime/wfengine/workflow.go

```diff
@@ -213,6 +216,70 @@ func (wf *workflowActor) getWorkflowMetadata(ctx context.Context, actorID string
 	return metadata, nil
 }
 
+// This method purges all the completed activity data from a workflow associated with the given actorID
+func (wf *workflowActor) purgeWorkflowState(ctx context.Context, actorID string) error {
+	state, exists, err := wf.loadInternalState(ctx, actorID)
+	if err != nil {
+		return err
+	} else if !exists {
+		return api.ErrInstanceNotFound
+	}
+
+	runtimeState := getRuntimeState(actorID, state)
+	if !runtimeState.IsCompleted() {
+		return api.ErrNotCompleted
+	}
+
+	// Loop through the history of the loaded workflow state and schedule purge calls for all the activities that have been completed
+
+	// RRL TODO: Looping through history and invoking each actor once. They are not in the same transaction. Eqch transaction happens as its own call
```

Comment:
As per the initial review during our 1:1 earlier today, this needs to change - although I admit I don't _fully_ understand how this works so I can't tell you exactly how to fix this yet 🙃

However, the problem is that the "purge" needs to happen in an atomic way. We discussed this and you explained that you need to delete 3 things:

- Each activity (this can be more than one)
- The history object
- The inbox object

These all have to be removed _at once_. We do that in Dapr by using the transactional state stores and performing a "transaction" (which are just operations performed atomically as a batch), but here, each activity is removed separately, in its own "transaction" (with a single opeation). Each activity is removed by the actor that you're invoking within the loop, and an error in the loop causes the entire method to exit, which means that _some_ state will have been removed, but not all.

Reply:
Hi @ItalyPaleAle. The state exists across potentially multiple actors, and AFAIK, we don't support transactions across multiple actors. We instead go for the next best thing, which is to do independent purge operations in a particular order so that failures that happen midway can be retried until there is finally a successful completion. This is why, for example, we delete state for each activity first before deleting the history (without the history, we wouldn't know which actors need to be deleted). Does that make sense? Let me know if you feel we need to meet and discuss this more in depth.

Reply:
I see. But what happens if there's a crash while the process is not complete?

Reply:
The crash will leave you in a partially deleted state. However, the workflow will still be visible from the GetWorkflow management APIs, and re-issuing the purge call should succeed.